### PR TITLE
Bounding Box Create Transformed method added, FamilyInstance null check added

### LINKED ISSUE #1258: Bounding box create transformed method
#### Description:

As in the title.

### PR DESCRIPTION
**Pull request by michal-pekacki**
### Issues addressed by this PR

Closes #1258 

- added `CreateTransformed` method for creating transformed bounding box
- null check added to `FamilyInstance`

### CODE REVIEW

File: Revit_Core_Engine/Compute/CreateTransformed.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Return transformed bounding box that is created from transformed corner points.")]
+        [Input("bbox", "Bounding box to transform.")]
+        [Input("transform", "Transformation of the bounding box.")]
+        [Output("bbox", "Transformed bounding box.")]
+        public static BoundingBoxXYZ CreateTransformed(this BoundingBoxXYZ bbox, Transform transform)
```
Comment (pawelbaran) [inline Revit_Core_Engine/Compute/CreateTransformed.cs:41]: From what I understand, you are not creating a transformed `BoundingBoxXYZ` here, but rather take its corner points and transform them, then found bounds of this transformed set? If so, then I would consider renaming the method to something around `Query.BoundsOfTransformed`.

File: Revit_Core_Engine/Compute/CreateTransformed.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Return transformed bounding box that is created from transformed corner points.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Compute/CreateTransformed.cs:37]: You are not returning a transformed bounding box, but a bounding box in global coordinate system encompassing transformed control points. This is a serious difference I think.

File: Revit_Core_Engine/Compute/CreateTransformed.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            XYZ pt7 = new XYZ(maxX, minY, maxZ);
+            XYZ pt8 = new XYZ(maxX, maxY, maxZ);
+
+            List<XYZ> points = new List<XYZ>() { pt1, pt2, pt3, pt4, pt5, pt6, pt7, pt8 };
+            List<XYZ> transPoints = points.Select(x => transform.OfPoint(x)).ToList();
+
+            double transMinX = transPoints.OrderBy(x => x.X).Select(x => x.X).FirstOrDefault();
+            double transMinY = transPoints.OrderBy(x => x.Y).Select(x => x.Y).FirstOrDefault();
+            double transMinZ = transPoints.OrderBy(x => x.Z).Select(x => x.Z).FirstOrDefault();
+            double transMaxX = transPoints.OrderBy(x => x.X).Reverse().Select(x => x.X).FirstOrDefault();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Compute/CreateTransformed.cs:79]: `OrderByDescending`

File: Revit_Core_Engine/Compute/CreateTransformed.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            XYZ pt4 = new XYZ(minX, maxY, maxZ);
+            XYZ pt5 = new XYZ(maxX, minY, minZ);
+            XYZ pt6 = new XYZ(maxX, maxY, minZ);
+            XYZ pt7 = new XYZ(maxX, minY, maxZ);
+            XYZ pt8 = new XYZ(maxX, maxY, maxZ);
+
+            List<XYZ> points = new List<XYZ>() { pt1, pt2, pt3, pt4, pt5, pt6, pt7, pt8 };
+            List<XYZ> transPoints = points.Select(x => transform.OfPoint(x)).ToList();
+
+            double transMinX = transPoints.OrderBy(x => x.X).Select(x => x.X).FirstOrDefault();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Compute/CreateTransformed.cs:76]: `OrDefault` seems to bring nothing here?
```suggestion
            double transMinX = transPoints.OrderBy(x => x.X).First().X;
```

File: Revit_Core_Engine/Create/FamilyInstance/FamilyInstance.cs [hunk 5]
```diff
@@ -264,6 +264,9 @@ private static FamilyInstance FamilyInstance_OneLevelBased(Document document, Fa
 
             //Place family instance
             FamilyInstance familyInstance = document.Create.NewUnionFitting(conn1, conn2);
+            if (familyInstance == null)
+                return null;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Create/FamilyInstance/FamilyInstance.cs:268]: Isn't it worth an error here?

File: Revit_Core_Engine/Compute/CreateTransformed.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        [Description("Return transformed bounding box that is created from transformed corner points.")]
+        [Input("bbox", "Bounding box to transform.")]
+        [Input("transform", "Transformation of the bounding box.")]
+        [Output("bbox", "Transformed bounding box.")]
+        public static BoundingBoxXYZ CreateTransformed(this BoundingBoxXYZ bbox, Transform transform)
+        {
+            if (bbox == null)
+                return null;
+
+            BoundingBoxXYZ newBBox = new BoundingBoxXYZ() { Min = bbox.Min, Max = bbox.Max };
```
Comment (pawelbaran) [inline Revit_Core_Engine/Compute/CreateTransformed.cs:46]: What about `bbox.Transform`? Shouldn't you do something like below?
```suggestion
            BoundingBoxXYZ newBBox = new BoundingBoxXYZ() { Min = bbox.Transform.OfPoint(bbox.Min), Max = bbox.Transform.OfPoint(bbox.Max) };
```

...with necessary null checks on transform ofc.

**Review by pawelbaran — CHANGES_REQUESTED** [review]

File: Revit_Core_Engine/Compute/CreateTransformed.cs (hunk 1 above)
Comment (michal-pekacki) [reply Revit_Core_Engine/Compute/CreateTransformed.cs:41]: Right, it returns the bounds after transform, not the transformed bounding box.

File: Revit_Core_Engine/Query/BoundsOfTransformed.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Return bounding box in global coordinate system that is created from transformed corner points.")]
+        [Input("bbox", "Bounding box of control points to transform.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/BoundsOfTransformed.cs:38]:
```suggestion
        [Input("bbox", "Bounding box with control points to transform.")]
```

File: Revit_Core_Engine/Query/BoundsOfTransformed.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        [Input("transform", "Transformation of the bounding box points.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/BoundsOfTransformed.cs:39]:
```suggestion
        [Input("transform", "Transformation to apply to the corners of the input bounding box.")]
```

File: Revit_Core_Engine/Create/FamilyInstance/FamilyInstance.cs [hunk 9]
```diff
@@ -264,6 +264,13 @@ private static FamilyInstance FamilyInstance_OneLevelBased(Document document, Fa
 
             //Place family instance
             FamilyInstance familyInstance = document.Create.NewUnionFitting(conn1, conn2);
+            if (familyInstance == null)
+            {
+                BH.Engine.Base.Compute.RecordError("Element could not be created due to unexpected errors.");
```
Comment (pawelbaran) [inline Revit_Core_Engine/Create/FamilyInstance/FamilyInstance.cs:269]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Element could not be created due to an internal error.");
```

**Comment by michal-pekacki** [comment]
@BHoMBot check required

**Comment by michal-pekacki** [comment]
@BHoMBot check required

**Comment by michal-pekacki** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**